[ews-build.webkit.org] Add Python hooks for results.webkit.org results-summary endpoint

JonWBedard opened the pull request:
#### 808a499669d0be1ed31fd49e69eaa948d5b0d4d5
<pre>
[ews-build.webkit.org] Add Python hooks for results.webkit.org results-summary endpoint
<a href="https://bugs.webkit.org/show_bug.cgi?id=247489">https://bugs.webkit.org/show_bug.cgi?id=247489</a>
rdar://101960268

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/results_db.py: Added.
(ResultsDatabase.get_results_summary): Return summary endpoint data.
(ResultsDatabase.is_test_expected_to): Compare summary endpoint data to a
threshold and return a binary result.
(ResultsDatabase.main): Simple command line interface for local verification
of EWS behavior.

Canonical link: <a href="https://commits.webkit.org/256346@main">https://commits.webkit.org/256346@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b277feb59a8a11decc19f6d4f36f49a908da490

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95499 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4767 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28552 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105081 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/165344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99486 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4794 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33501 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87870 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101163 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82112 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30576 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/87304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73413 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/39261 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/94476 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20159 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4388 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/40958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/42946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39406 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python3
+
+# Copyright (C) 2020, 2021 Apple Inc. All rights reserved.
```
Nit: 2022

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+
+
+class ResultsDatabase(object):
+    URL = 'https://results.webkit.org'
```
Nit: maybe `HOSTNAME` instead of `URL`

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+
+class ResultsDatabase(object):
+    URL = 'https://results.webkit.org'
+    SUITE = 'layout-tests'
```
Maybe add a TODO to support other suites as well (like api-tests).

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+    THRESHOLD = 10
+
+    @classmethod
+    def summary(cls, test, commit=None, configuration=None):
```
would be good to have more descriptive names, e.g.: `summary ` => `get_test_results_summary`

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+        return response.json()
+
+    @classmethod
+    def expected(cls, test, type=None, commit=None, configuration=None, log=False):
```
Ditto, `expected` => `is_test_expected_to_fail`
Nit, maybe `log` => `verbose`
`commit` => `commit_identifier / commit_id / change_id/ commit_ref`
what's `type`? probably better to have more descriptive name.

JonWBedard replied:
I'm changing it to `result_type`, it's an optional argument, basically would let you differentiate between `timeout`/`crash`/`text/image/audio` failures

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+            if not data:
+                print('    No historic data found for query')
+        if not data:
+            return False
```
currently this method return False for two scenario, 
1) when the test is not expected to fail, and 
2) when there is no data found.

Probably, better to return integer, and then -1 to indicate no data, so that the callers can act accordingly.

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+        if not data:
+            return False
+        if type:
+            return cls.THRESHOLD < data.get(type.lower(), 0)
```
Nit: for readability, it might be better to reverse the order: 
`return data.get(type.lower(), 0) > cls.THRESHOLD`

Also, better to return something like - 1 when data is not found.

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+            return False
+        if type:
+            return cls.THRESHOLD < data.get(type.lower(), 0)
+        return cls.THRESHOLD < 100 - (data.get('pass', 0) + data.get('warning', 0))
```
Ditto

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+class ResultsDatabase(object):
+    URL = 'https://results.webkit.org'
+    SUITE = 'layout-tests'
+    THRESHOLD = 10
```
Better to have more descriptive name indicating threshold for what exactly.

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+            '{}/api/results-summary/{}/{}'.format(cls.URL, cls.SUITE, test),
+            params=params,
+        )
+        if response.status_code != 200:
```
would be good to log something here.

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python3
```
Would be good to keep the script name generic, like `results_db.py` or something like that. Currently it handles the case of pre-existing failures, but we would soon want to add the case for determining if a test is flaky failure (as ews encounters a lot of flaky tests). Similarly, there might be more functionality needed for results db interactions.

JonWBedard replied:
This actually does the flaky case already. The endpoint we're hitting gives weights based on how far from the specified commit results are coming from.

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+
+    @classmethod
+    def expected(cls, test, type=None, commit=None, configuration=None, log=False):
+        data = cls.summary(test, commit=commit, configuration=configuration)
```
what happens if the passed commit sha /identifier doesn't yield any data on results db, does results db produce results based on nearby commits?

JonWBedard replied:
Yes, even if the tests did give a result on the specific commit, we still take into account commits around it. There is an algorithm which basically applies weights to each result based on how far it is from the specified commit, closer being given more weight.

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+        parsed = parser.parse_args(args)
+
+        configuration = dict()
+        for key in ['style']:
```
Nit: might want to add more keys here, although can be done in subsequent PRs as well.

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+
+    @classmethod
+    def summary(cls, test, commit=None, configuration=None):
+        params = dict(**(configuration or dict()))
```
should have some basic validation if configuration passed is valid or it contains invalid keys. But that can be done in subsequent PRs.

aj062 reviewed (approved):
Logic looks good. Couple of comments, mostly related to naming.

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+        if commit:
+            params['ref'] = commit
+        response = requests.get(
+            '{}/api/results-summary/{}/{}'.format(cls.URL, cls.SUITE, test),
```
Feel free to use f-string here (since ews code is Python 3 only).

aj062 commented on Tools/CISupport/ews-build/is_failing.py:
```diff
@@ -0,0 +1,126 @@
+        )
+        if response.status_code != 200:
+            return None
+        return response.json()
```
would be good to wrap `.json()` call in a try except block, so that it doesn't raise an exception while running on ews.

JonWBedard committed:
[ews-build.webkit.org] Add Python hooks for results.webkit.org results-summary endpoint
https://bugs.webkit.org/show_bug.cgi?id=247489
rdar://101960268

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/results_db.py: Added.
(ResultsDatabase.get_results_summary): Return summary endpoint data.
(ResultsDatabase.is_test_expected_to): Compare summary endpoint data to a
threshold and return a binary result.
(ResultsDatabase.main): Simple command line interface for local verification
of EWS behavior.

Canonical link: https://commits.webkit.org/256346@main
